File: usaco/training/section_6.1/rectbarn.py
"""
ID: huangch7
LANG: PYTHON3
TASK: rectbarn
"""


task_name = "rectbarn"
INF = 100000000


def get_list_from_line(dtype=int):
    return [dtype(x) for x in fin.readline().strip().split()]


def get_strs_from_line():
    return fin.readline().strip().split()



fin = open(f'{task_name}.in', 'r')
fout = open(f'{task_name}.out', 'w')

R, C, P = get_list_from_line()

field = [[0] * (C + 1) for _ in range(R + 1)]
for _ in range(P):
    x, y = get_list_from_line()
    field[x][y] = 1

# for i in range(R + 1):
#     print(field[i])

height = [[1] * (C + 2) for _ in range(R + 1)]
for i in range(1, R + 1):
    for j in range(1, C + 1):
        if field[i][j] == 1:
            height[i][j] = 0
        elif i > 1:
            height[i][j] = height[i - 1][j] + 1
    height[i][-1] = 0

best = 0
for i in range(1, R + 1):
#     print(height[i])
    stack = []
    for j in range(1, C + 2):
        idx, h = None, None
        while len(stack) and stack[-1][1] > height[i][j]:
            idx, h = stack.pop()
            area = (j - idx) * h
            best = max(best, area)
#             print('debug', j, idx, h)
            
        if idx is None:
            stack.append((j, height[i][j]))
        else:
            stack.append((idx, height[i][j]))
#         print(j, stack, best)
    
fout.write(f"{best}\n")
            

fout.close()

File: usaco/training/section_6.1/rectbarn.cpp
/*
ID: huangch7
LANG: C++
TASK: rectbarn
*/

#include <iostream>
#include <fstream>
#include <stack>
#include <algorithm>
#include <unordered_set>
using namespace std;

const int MAX_R = 3000;
const int MAX_C = 3000;
const int INF = 100000000;

ifstream fin("rectbarn.in");
ofstream fout("rectbarn.out");

int R, C, P;

// int field[MAX_R + 1][MAX_C + 1];
// int height[MAX_R + 1][MAX_C + 2];

int get_int() {
    string line;
    getline(fin, line);
    return stoi(line);
}

int main() {
  int R, C, P;
  fin >> R >> C >> P;

  //  int** field = new int*[R + 1];
  //  int** height = new int*[R + 1];
  int** height = new int*[2];  
  for (int i = 0; i < 2; ++i) {
    //    field[i] = new int[C + 2];
    height[i] = new int[C + 2];
  }  
  /*
    for (int i = 0; i <= R; ++i) {
        for (int j = 0; j <= C; ++j) {
            field[i][j] = 0;
        }
    }

    for (int i = 0; i < P; ++i) {
      int x, y;
      fin >> x >> y;
        field[x][y] = 1;
    }
  */
  unordered_set<int> dmg;
  for (int i = 0; i < P; ++i) {
    int x, y;
    fin >> x >> y;
    dmg.insert(x * 5000 + y);
  }
  /*    for (int i = 0; i < 2; ++i) {
        for (int j = 0; j <= C; ++j) {
            height[i][j] = 1;
        }
    }
  */
    int best = 0;
    for (int i = 1; i <= R; ++i) {
      for(int j = 0; j <= C; ++j){
        height[i % 2][j] = 1;
      }
      height[i % 2][C + 1] = 0;      
        for (int j = 1; j <= C; ++j) {
          //            if (field[i][j] == 1) {
          if (dmg.find(i * 5000 + j) != dmg.end()){
            height[i % 2][j] = 0;
          } else if (i > 1) {
            height[i % 2][j] = height[(i - 1) % 2][j] + 1;
          }
        }

        //    }


        //    for (int i = 1; i <= R; ++i) {
        stack<pair<int, int> > st;
        for (int j = 1; j <= C + 1; ++j) {
            int idx = -1, h = -1;
            while (!st.empty() && st.top().second > height[i % 2][j]) {
                idx = st.top().first;
                h = st.top().second;
                st.pop();
                int area = (j - idx) * h;
                best = max(best, area);
            }
            if (idx == -1) {
                st.push({j, height[i % 2][j]});
            } else {
                st.push({idx, height[i % 2][j]});
            }
        }
    }

    fout << best << endl;

  for (int i = 0; i < 2; ++i) {
    //    delete[] field[i];
    delete[] height[i];
  }
  //  delete [] field;
  delete [] height;

    return 0;
}
